### PR TITLE
Use CSPLayout in Level 2 and Level 3

### LINKED ISSUE #3386: Optim level 2 and 3 modifiy circuits that match topology
### Information

- **Qiskit Terra version**: master 
- **Python version**:
- **Operating system**:

### What is the current behavior?
The following circuit fits Vigo-like topologies exactly:

Screen Shot 2019-11-04 at 2 35 42 PM

But optimization level 3 gives:

Screen Shot 2019-11-04 at 2 35 56 PM

Where as the default gives the expected original circuit (save for H -> U2):

Screen Shot 2019-11-04 at 2 36 13 PM

### Steps to reproduce the problem
```python

qc = QuantumCircuit(5)
qc.h(3)
qc.cx([3, 3, 1, 1], [1, 4, 2, 0])
qc.measure_all()
new_qc = transpile(qc, backend, optimization_level=3)
new_qc.draw(output='mpl')
```

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #3386

Before trying `DenseLayout`, try `CSPLayout`.

### CODE REVIEW

**Comment by 1ucian0** [comment]
on top of PR #3043 , therefore `on hold`

**Comment by ajavadia** [comment]
@1ucian0 needs a rebase

**Comment by kdk** [comment]
Out of curiosity, I quickly ran `CSPLayout` over some random CX circuits of varying width/depth across the fake backends. It looks like there are a some cases (shallow, wide circuits on wide backends) where the runtime will be on the order of several seconds. 

image

(The subplots here cover circuits composed of 1) gates entirely from the coupling map, 2) coupling map gates + 1 random gate outside the coupling map, 3/4/5) 20, 50, 80 percent gates outside the coupling map.)


image

I only included the <= 20q devices, but tried on `FakeRochester` this morning and ran into a handful of cases that spun for at least several minutes.

On the plus side, the circuits for which `CSPLayout` seems to be the slowest are in general shallow, so the benefits of a compatible layout are large and remainder of the transpilation should be comparatively fast, but I don't think we can put it in the default pipeline without something to guard the runtime (maybe a timeout, or limitation on device size).

**Comment by 1ucian0** [comment]
Awesome analysis!

Shallow circuits mean wider trees, resulting in more walking without
triggering backtracking. So kinda makes sense.

Maybe time out is a good approach for different levels?

**Comment by 1ucian0** [comment]
`on hold` until we limit how long the pass runs (PR https://github.com/Qiskit/qiskit-terra/pull/3515)

**Comment by 1ucian0** [comment]
#3515 merged. Removing on hold

**Comment by 1ucian0** [comment]
This PR adds CSPLayout with the following parameters:
 - level 2:  `call_limit=1000, time_limit=10` (default parameters)
 - level 3:  `call_limit=10000, time_limit=60`

File: releasenotes/notes/csplayout-fb9deb7fc133acfd.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+---
```
Comment (kdk) [inline releasenotes/notes/csplayout-fb9deb7fc133acfd.yaml]: I think this is from when `CSPLayout` was introduced. Can you update it to note that it is being included in levels 2 and 3 with default timeouts/call limits?

Comment (1ucian0) [reply releasenotes/notes/csplayout-fb9deb7fc133acfd.yaml]: Indeed. When the release was done this file was moved to 0.11.  Removing and adding new release notes.

**Pull request merged**